### ISSUE TITLE
Try TS's new `import()` w/ the esri types

### ISSUE THREAD

**Issue by tomwayson** [Esri/esri-loader#119]
I suspect that TS's new syntax to `import()` types could be just what esri-loader has been looking for in terms of a clean story for how to use the esri types.

https://davidea.st/articles/typescript-2-9-import-types

**Comment by andygup** [comment]
Nice, I also was thinking that! I suspect that we'll want to add `// comments` to provide context that we aren't importing modules as well as for those who aren't at `TS 2.9` yet. 

We'll probably want to update related repos, as well: `angular-cli-esri-map`, `ionic2-esri-map` etc etc

**Comment by davewilton** [comment]
Tom, 

I just had a brief look at this as for a solution in project rather than how to fix the loader. Maybe I'm missing something or its to do with the way the esri types are constructed. But what you get back is the interface for map view and therefore not the constructor:

```
initializeMap = async (container: HTMLElement) => {
    loadCss('https://js.arcgis.com/4.10/esri/css/main.css');
    // const [MapView, WebMap] = await loadModules(['esri/views/MapView', 'esri/WebMap']);
    const MapView = await this.getMapView();
    const WebMap = await this.getMapView();
    // then we load a web map fromy an id
    const webmap = new WebMap({
      portalItem: {
        id: this.webmapid
      }
    });
    // and we show that map in a container w/ id #viewDiv
    const view = new MapView({
      map: webmap,
      container: container as HTMLDivElement
    });

    view.when().then(() => {
      console.log('map is ready');
      this._invalidator();
    });
  }

  async getMapView(): Promise<import('esri/views/MapView')> {
    return await loadModules(['esri/views/MapView']) as any;
  }
  async WebMap(): Promise<import('esri/WebMap')> {
    return await loadModules(['esri/WebMap']) as any;
  }
```
```
Cannot use 'new' with an expression whose type lacks a call or construct signature.
```

Am I missing something obvious here?

**Comment by tomwayson** [comment]
I never got around to trying this, but I think you're close. Each of your calls to `loadModules()` is returning an array, not the class, so something like `return loadModules(['esri/views/MapView'])[0]` might work.

**Comment by tomwayson** [comment]
Also, FYI - I doubt this is related, but I see you're calling `this.getMapView()` 2x, I think you want `const WebMap = await this.WebMap();`

**Comment by davewilton** [comment]
Sorry Tom, I was playing around a lot. Ignoring my errors there are still compile time issues. It's the same issue as if you try to type the items using the global esri namespace e.g.

```
 const [m, w] = await loadModules(['esri/views/MapView', 'esri/WebMap']);
    const MapView: __esri.MapView  = m;
    const WebMap: __esri.WebMap = w;
```

Its something to do with the  interface being returned rather than the webMapConstructor.

The code actually works, but we are getting compile time issues.

Here you can see them in a sandbox, the app works but there are TS compile issues (please ignore hooky react code)

https://codesandbox.io/embed/xv8mw2890w

![image](https://user-images.githubusercontent.com/2278405/54962572-79a65a00-4fb9-11e9-92f3-b654bc782809.png)

**Comment by davewilton** [comment]
Nevermind @tomwayson I solved it (add typeof). 

Any thoughts on how we can get this into the project> My initial thought is really basic and I'm sure there is a much better way... but if we have a loadModule singular we could have a large number of overrides:

```
export declare function loadModule(module: string, loadScriptOptions?: ILoadScriptOptions): Promise<any>;
export declare function loadModules(modules: string[], loadScriptOptions?: ILoadScriptOptions): Promise<any[]>;
export declare function loadModule(module: "esri/WebMap", loadScriptOptions?: ILoadScriptOptions): Promise<typeof import('esri/WebMap')>;
export declare function loadModule(module: "esri/views/MapView", loadScriptOptions?: ILoadScriptOptions): Promise<typeof import('esri/views/MapView')>;
```

Gives the added advantage of hints in import
![image](https://user-images.githubusercontent.com/2278405/54963466-0a326980-4fbd-11e9-92f9-e97ad2f0d83f.png)

Obviously this involves getting a list of all the esri modules as a string and maintaining the list (although users can always import items we haven't added through the first option)

**Comment by tomwayson** [comment]
@davewilton that's great that you got it to work!

I think whenever possible it's best to load more than one module at a time, so I'm trying to find a way to apply what you've done to a function that would return an array of modules but I can't get it to work:

```ts
  type MyModules = [typeof import('esri/WebMap'), typeof import('esri/views/MapView')];
  function loadMyModules(): Promise<MyModules> {
    return loadModules(['esri/WebMap', 'esri/views/MapView']) as MyModules;
  }
```

I get:

![image](https://user-images.githubusercontent.com/662944/55095035-b420fd00-5074-11e9-8de2-355bebb45149.png)

Were you ever able to get something like that to work?

**Comment by tomwayson** [comment]
Actually you don't even need the wrapping function:

```ts
  type MyModules = [typeof import('esri/WebMap'), typeof import('esri/views/MapView')];
  (loadModules(['esri/WebMap', 'esri/views/MapView']) as Promise<MyModules>)
  .then(([WebMap, MapView]) => {
    const map = new WebMap({
      portalItem: {
        id: 'abdef'
      }
    });
    // and we show that map in a container w/ id #viewDiv
    const view = new MapView({
      map,
      container: `viewDiv`
    });
  });
```

**Comment by davewilton** [comment]
Looks good @tomwayson, I was worried about potential performance issues which you would understand better. 

So we are proposing a change to the readme rather than changing the library? I can't think of a way we would integrate this into the library other than the one I proposed. Add a "typescript" section to the readme? I could draft something this week if you like

**Comment by tomwayson** [comment]
@davewilton that's fantastic!

I actually tried whipping up a codesandbox to demonstrate this earlier, but it crashed Chrome into an un-recoverable state, so I didn't finish. Once I get a chance to reboot, I'll take a look at that in more detail.

> So we are proposing a change to the readme rather than changing the library?

Exactly.

There is already a README section on using [ArcGIS Types](https://github.com/Esri/esri-loader/#arcgis-types). I propose adding a section about using TS `import()` for > TS 2.9 and linking to your codesandbox.

It'd be great if you would PR that^^^!

I'm going fork your sandbox and try and get it working for 3.x. If I can, I'll add on to your PR w/ a link to that.

**Comment by davewilton** [comment]
@tomwayson I created a V3x sandbox too and a pull request. Its a bit dense with the 3x/4x and <2.9 >2.9 to get in there. Hope its ok

**Comment by tomwayson** [comment]
resolved in #164
